### ISSUE TITLE
Unique Job ID

### ISSUE THREAD

**Issue by krazyjakee** [reswifq/reswifq#5]
Hi!

Very happy with our fancy job queue, thank you.

One thing, we need to be able to ensure jobs are unique. Other solutions have a JobID (string) to ensure unique jobs. Does reswifq have such a feature? I couldn't see anything in the documentation.

**Comment by valeriomazzeo** [comment]
Hi @krazyjakee.

Each `Job` is managed by a `JobBox` container and it has a unique identifier (see https://github.com/reswifq/reswifq/blob/master/Sources/Reswifq/Job/JobBox.swift#L29).

**Comment by krazyjakee** [comment]
I would like to set this value myself so I can ensure the same job is not added twice.

So 2 steps:
- [ ] Allow the identifier to be set as a property of the Job
- [ ] Ensure the identifier does not exist in the queue before adding it to the queue

I guess also
- [ ] Add a test

**Comment by valeriomazzeo** [comment]
>I would like to set this value myself so I can ensure the same job is not added twice.

I am not sure that is `Reswifq` responsibility to check for jobs to be unique. It looks to me like is something very specific to the application logic itself.

The current implementation ensure that a `Job` is enqueued for each `enqueue` call.
Imagine the following scenario where you have a `SendMail` job, and you want to schedule it, every 5 minutes for the next couple of hours (just as an example).

```swift
let sendMail: SendMailJob()
let count = Int(60 * 2 / 5)
for index in 1..count {
  reswifq.enqueue(sendMail, scheduleAt: Date(timeIntervalSinceNow: index * 300)
}
```

With your proposed change to achieve the same:
```swift
let count = Int(60 * 2 / 5)
for index in 1..count {
  reswifq.enqueue(SendMailJob(), scheduleAt: Date(timeIntervalSinceNow: index * 300)
}
```

It doesn't look to me like there is a right or wrong, and I think by design the current implementation is the most flexible one.

Consider the following:

 * In theory, it could be done in `MemQueue`, but my first thought about the default behaviour if another job with the same id is enqueued, would be to move it to the back to the queue. It would be interesting to know what other queue processing system do in this scenario (RabbitMQ, sidekiq, etc).
Main point being, as you can see we already have a two different opinions on how unique jobs should behave.

* I am not sure how this is possible in `Reswifq`. Jobs are stored in a list using `LPUSH` so the only way I can think of, to check the identifier, is to fetch, decode, and match the id of every single job for each enqueue operation. It doesn't make much sense.

I am happy to talk about it, but so far I believe that this kind of logic is very application specific and if you require for the same job instance (or any other criteria specific to your needs) to not be enqueued more than once, then it should be implemented in your application. 

Also remember, `Job` is a protocol and you can extend it as you need.

**Comment by krazyjakee** [comment]
The biggest use case is when the server stops for any reason, it can happily pick up where it left off without worrying what is in the queue and what isn't. What you're describing about having to parse the id of each job and match it, that is what anyone who needs unique jobs will have to do anyway with the current system. It would be better to have that as a convenient optional feature of the library.

It would be better to hold the job ID in the redis key, then it can be easily matched using the redis index. The rest of the data can be held as a value inside that key.
Something like `queue:state:priority:jobId`

sidekiq enterprise offer unique jobs: https://github.com/mperham/sidekiq/wiki/Ent-Unique-Jobs
sidekiq has a heavily used extension to achieve this too: https://github.com/mhenrixon/sidekiq-unique-jobs
They simply ignore the queued item when it's created if it already exists. If you add a unique job to a queue and it's behaviour is to add that to the back of the queue, then it's not a unique job. If you want to have that kind of functionality that would be a numerical priority which is a different feature.

RabbitMQ has no solution for this [much to the dismay of those who need it](https://stackoverflow.com/a/13148079/1013097).

I have used [Bull queue](https://github.com/OptimalBits/bull) extensively and they [allow you to set a jobId](https://github.com/OptimalBits/bull/blob/master/REFERENCE.md#queueadd) and then append the jobId to the redis key to allow for quick lookup.

**Comment by valeriomazzeo** [comment]
>The biggest use case is when the server stops for any reason, it can happily pick up where it left off without worrying what is in the queue and what isn't.

It would be interesting to have a more detailed example because there must be some detail I am missing about this.

>It would be better to hold the job ID in the redis key, then it can be easily matched using the redis index. The rest of the data can be held as a value inside that key.
Something like queue:state:priority:jobId

Yes, unfortunately the solution I mentioned is not really an option. On the other hand, moving away from using the list means we can’t use RPOPLPUSH anymore and that open a whole new can of worms with having to use transactions. Although, if sorted may lead to be able to implement queue priorities.

I remember considering this approach but then for some reason I moved away from it.

I could definitely reconsider it, but I am afraid I won’t have the time anytime soon for such a big change.

If you are willing to make a pull request I would be happy to review it and to help.

**Comment by krazyjakee** [comment]
After some research I understand your points and of course, why you chose to use lists. A sorted set would enable job uniqueness and priority but, as you say, it would require a transaction to RPOPLPUSH.

You can see this key function in sidekiq where this transaction takes place: https://github.com/mperham/sidekiq/blob/d61a7477337a7ba1ef522d80bfe28b3b3c885414/lib/sidekiq/client.rb#L191-L207

We repeatedly pull data from a source then add it to a queue every 24 hours. It's many hundreds of thousands of items and on the next pull, it may inadvertently add the same items into the queue. It's very important this data is not aggregated twice which is why we need to ensure uniqueness at all times. We could store an array in application memory but if it went down for any reason we would lose that list.

We started modifying the code in our fork but sadly our project is too time sensitive to make such a large change. As you say, we can work-around this in our application. We will use a redis sorted list that holds the queue keys and make sure it does not add duplicate job items with reswifq.

We will use this library over and over again in the future so a pull-request is coming but probably in Q1 next year.

Thanks so much for your help on this, any further input you have would be greatly appreciated.

**Comment by valeriomazzeo** [comment]
>We could store an array in application memory but if it went down for any reason we would lose that list.

Correct, I definitely won't advise for this solution. If you persist your queue (for example using Redis), only the persistent store is able to let you know about already existing enqueued jobs.

The whole Reswifq system has been architected to be easily extendable. Thanks to the `Queue` protocol, there is no hard dependency between a specific concrete queue implementation and the rest of the system. This means that you can create your own queue class and still be able to plug it and use it into the existing processes (see `Reswifc`, `Worker`, etc). `MemQueue` is a good example of this.

My recommendation is to not be necessarily tied to `Redis`, you could in fact create a `Reswifq` pluggable `Queue` which is implemented on a different persistence service that suits more your needs.

>We will use this library over and over again in the future so a pull-request is coming but probably in Q1 next year.
>
>Thanks so much for your help on this, any further input you have would be greatly appreciated.

I am really looking forward to see what people can do with Reswifq.

Wish you the best with your project :)
